### PR TITLE
[Generator] Part 2 of Them!

### PR DESCRIPTION
**Pull request by NycuRO**
Hello guys.
As you know, in the past, I and @Creeperface01 have added improvements to the generator. Well, those improvements are very few besides what I have in my head. I would like to add more stuff, all biomes and lots of stuff besides!

Well, I said I'd come back with change and here I came!

/* 
**Changes:**
//1. [ ] Implement Water Color.
//2. [ ] Add all Biomes. (Generate + File)
//3. [ ] Add more Populators.
4. [X] Remove BIOMES from cn.nukkit.level.generator.normal.Normal and use instead from cn.nukkit.level.generator.normal.biome.Biome.
5. [X] Use Mountains and SmallMountains class as one abstract and extends ExtremeHills biomes.
//6. [ ] Implement Nether fully functional.
//7. [ ] Replace Water with ICE in ICE_FLATS + TAIGA.
//8. [ ] Fix Caves.
//9. [ ] Fix all Warnings for get fuck off these 'errors':
```
20:20:03 [DEBUG] Level "world" took 52.0ms, setting tick rate to 2 ticks
20:20:03 [DEBUG] Raising level "world" tick rate to 1 ticks
```
//10. [ ] Fix Snows on Trees in ICE_FLATS. + Get More Colors to Leaves (White) 😄 
11. [X] Fix World what isn't NETHER/END and get other Sky.
12. Others.
*/
**These branch IS TESTED BUT NOT HAVE ALREADY ALL.**

**All can download .jar from there: https://circleci.com/gh/NycuRO/Nukkit for test JAR**
**ATENTION, BRANCH MASTER**

### CODE REVIEW

**Comment by NycuRO** [comment]
+ trying fix non-generated Ice Block instanceof Water in ICE_FLATS + TAIGA biomes.

**Comment by NycuRO** [comment]
Tommorow i add and Populators and others.

**Comment by NycuRO** [comment]
In these PR i add fully nether.

**Comment by NycuRO** [comment]
image

Colored water works.

**Comment by NycuRO** [comment]
image

I'm watched color biomes works, but it's one problem with Beach Biome.  When them is generated, change color.
Need add one function if exist Beach Biome and water, get Biome color.

**Comment by vzlomka76** [comment]
Cool :+1:

**Comment by NycuRO** [comment]
+ testing Color water if exist BEACHES biomes

File: src/main/java/cn/nukkit/level/generator/Normal.java [hunk 1]
```diff
@@ -1,5 +1,6 @@
 package cn.nukkit.level.generator;
 
+import a.c.B;
```
Comment (NycuRO) [inline src/main/java/cn/nukkit/level/generator/Normal.java]: again that.........

**Comment by NycuRO** [comment]
image

Float values in lookUp get Biomes to more longest. Watch on Savanna 👍

**Comment by NycuRO** [comment]
8. [ ] Fix Caves.
9. [ ] Fix all Warnings for get fuck off these 'errors':
```
20:20:03 [DEBUG] Level "world" took 52.0ms, setting tick rate to 2 ticks
20:20:03 [DEBUG] Raising level "world" tick rate to 1 ticks
```

Added to TODO.

**Comment by NycuRO** [comment]
Bug:

image


Client send to ICE_FLATS on tree SNOW, but server get them off (Remove them) and get more lag on generation.

File: src/main/java/cn/nukkit/level/generator/Normal.java [hunk 2]
```diff
@@ -352,7 +353,8 @@ public void generateChunk(int chunkX, int chunkZ) {
                     if (geny <= bedrockDepth && (geny == 0 || nukkitRandom.nextRange(1, 5) == 1)) {
                         chunk.setBlock(genx, geny, genz, Block.BEDROCK);
                     } else if (geny > genyHeight) {
-                        if ((biome.getId() == Biome.ICE_FLATS || biome.getId() == Biome.TAIGA) && geny == seaHeight) {
+                        if ((biome.getId() == Biome.ICE_FLATS || biome.getId() == Biome.TAIGA) && geny == seaHeight && biome.getId() == Biome.BEACHES) {
```
Comment (NycuRO) [inline src/main/java/cn/nukkit/level/generator/Normal.java]: River not Beaches...

**Comment by NycuRO** [comment]
Nukkit rewrited generator "Normal" like that:
```
if (geny <= bedrockDepth && (geny == 0 || nukkitRandom.nextRange(1, 5) == 1)) {
                        chunk.setBlock(genx, geny, genz, Block.BEDROCK);
                    } else if (geny > genyHeight) {
                        if ((biome.getId() == Biome.ICE_FLATS || biome.getId() == Biome.TAIGA) && geny == seaHeight) {
                            // Trying fix ungenerated ice in These Biomes
                            chunk.setBlock(genx, geny, genz, Block.ICE);
                        } else {
                            chunk.setBlock(genx, geny, genz, Block.STILL_WATER);
                        }
                    } else {
                        chunk.setBlock(genx, geny, genz, Block.STONE);
                    }
```

Water is replaced by ICE if is ICE_FLATS and TAIGA biomes.
Yes correctly, but RIVER exist on them, like in "[1] Bug".
For fix them, need add RIVER biome. But this not work.
**Have anyone other solution?**

**Comment by NycuRO** [comment]
image


Ok so much bugs is created and by "Mountains"..

This mountains who added Nukkit by code is MUTATED Biomes, not Special Mountains and is generated with "Natural" biomes.

Need rewrite this.

**Comment by NycuRO** [comment]
image

Ok, my method for remediate Mountains works, but not fully. Need more test. 
**Changes:**
Before:
```java
if ((biome.getId() == Biome.ICE_FLATS || biome.getId() == Biome.TAIGA) && geny > seaHeight) {
```
After:
```java
if ((biome.getId() == Biome.ICE_FLATS || biome.getId() == Biome.TAIGA) && geny >= seaHeight) {
```

**Comment by NycuRO** [comment]
I have other ideea for Mountains. Now is good:

Need replace Mountains per biome with MUTATED_biome and all is good.

**Comment by NycuRO** [comment]
* Buket not Bukkit sorry :)))

**Comment by NycuRO** [comment]
image

Yey, portal works now! Good!

**Comment by NycuRO** [comment]
image

Breaks works!

**Comment by NycuRO** [comment]
Bug:

image

**Comment by NycuRO** [comment]
image

Added auto-generated nether world, but portal not work. Need fix.

**Comment by NycuRO** [comment]
Other bug:
Nether world is not "nether"

**Comment by SupremeMortal** [comment]
It would be better if you did all the testing on your computer then pushed the changes here.

**Comment by NycuRO** [comment]
@SupremeMortal If you say why don't compile, or you can help me with that, i compile on my computer.

image

**Comment by SupremeMortal** [comment]
Just search for a tutorial on how to compile with maven on Intellij. 
Doing `mvn idea:idea` with Maven installed in the Nukkit repo will setup the appropriate files but I don't use Idea so I can't tell you anything else.

**Comment by NycuRO** [comment]
Thanks @SupremeMortal

**Comment by NycuRO** [comment]
@SupremeMortal Not work. If i compile:

image

**Comment by larryTheCoder** [comment]
@NycuRO the language file is missing you need to download the [language files](https://github.com/Nukkit/Languages/). paste the file into your %PROJECT_FOLDER%\src\main\resources\lang

Note: %PROJECT_FOLDER% is your nukkit project path

**Comment by NycuRO** [comment]
@larryTheCoder Ok larry

**Comment by NycuRO** [comment]
image

Auto-Generated Nether World [X]

image

Teleportation [X].
Change Dimension Packet with Spawn Packet [X].

**TODO/Bugs:**

~~~~~~~~~~ 1 ~~~~~~~~~~~

![image](https://user-images.githubusercontent.com/13573011/29040033-40eb5162-7bb6-11e7-9353-83695f88392d.png)



~~~~~~~~~~ 2 ~~~~~~~~~~~

When you are teleported in Nether, portal is not created.

~~~~~~~~~ 3 ~~~~~~~~~~~~

I don't think this is bug, but when you are teleported, exist possibility to teleport you on Blocks. ( Distance Nether = Distance Normaly x 8 *WORK* and these probably create this "bug".
I'm used getSafeSpawn() ,but same.

~~~~~~~~ 4 ~~~~~~~~~~~~~

When you kill in nether:

![image](https://user-images.githubusercontent.com/13573011/29040155-b28817f6-7bb6-11e7-9f21-5b57d915b7f3.png)

~~~~~~~~ 5 ~~~~~~~~~~~~

![image](https://user-images.githubusercontent.com/13573011/29040200-caf61694-7bb6-11e7-9a74-8ae58871cfda.png)


**If anyone can help me to solve that, it's so best guy** 👍

**Comment by Creeperface01** [comment]
Also nether portal position in the nether is calculated from position of portal from overworld

**Comment by NycuRO** [comment]
@Creeperface01 
```java
this.nukkitRandom.setSeed(0xdeadbeef ^ (chunkX << 8) ^ chunkZ ^ this.level.getSeed());
```

**Comment by PikyCZ** [comment]
And end?

**Comment by NycuRO** [comment]
@PikyCZ End no. In part 3.

**Comment by larryTheCoder** [comment]
Let me do the end :D

**Comment by NycuRO** [comment]
image



Optimizations work so good 👍

**Comment by NycuRO** [comment]
image


I like this image so much <3

**Comment by NycuRO** [comment]
image


And i love mountains isn't to big and much shitty <3

**Comment by sero583** [comment]
@NycuRO Notice: Do not imlement generators on your opinon, implement they like are in real Minecraft but good job

**Comment by NycuRO** [comment]
@sero583 i know that. Code is different, actions is same.

**Comment by NycuRO** [comment]
@boy0001 This PR brokened your plugin (FAWE)

**Comment by SupremeMortal** [comment]
This hasn't been merged so mentioning the author of the plugin because it is incompatible seems irrelevant.

**Comment by NycuRO** [comment]
@SupremeMortal Yes, i know that. But is by setDimension and no changes now in there :D

**Comment by sero583** [comment]
@NycuRO This does not effects anything unfortunately

**Comment by NycuRO** [comment]
@sero583 Ok.

**Comment by sero583** [comment]
@NycuRO I mean I even call generateChunk & loadChunk (using if possible force method) but somehow I dont think it loads/generated, so maybe take a look how it gets done on loaded chunks?

**Comment by NycuRO** [comment]
@sero583 I don't do that. It's not recomanded.

**Comment by sero583** [comment]
@NycuRO with that I tried to explain not even after that it works.

Also btw on your PR this Exception occurred:
``19:24:47 [ALERT] java.lang.ArrayIndexOutOfBoundsException: -1                                     │·
        at cn.nukkit.level.format.generic.BaseChunk.getBlockLight(BaseChunk.java:135)             │·
        at cn.nukkit.level.Level.getBlockLightAt(Level.java:2258)                                 │·
        at cn.nukkit.level.Level.computeRemoveBlockLight(Level.java:1602)                         │·
        at cn.nukkit.level.Level.updateBlockLight(Level.java:1568)                                │·
        at cn.nukkit.level.Level.updateAllLight(Level.java:1531)                                  │·
        at cn.nukkit.level.Level.setBlock(Level.java:1678)                                        │·
        at cn.nukkit.block.Block.place(Block.java:677)                                            │·
        at cn.nukkit.level.Level.useItemOn(Level.java:2086)                                       │·
        at cn.nukkit.Player.handleDataPacket(Player.java:2338)                                    │·
        at cn.nukkit.network.Network.processPackets(Network.java:175)                             │·
        at cn.nukkit.network.Network.processBatch(Network.java:151)                               │·
        at cn.nukkit.Player.handleDataPacket(Player.java:2039)                                    │·
        at cn.nukkit.network.RakNetInterface.handleEncapsulated(RakNetInterface.java:157)         │·
        at cn.nukkit.raknet.server.ServerHandler.handlePacket(ServerHandler.java:123)             │·
        at cn.nukkit.network.RakNetInterface.process(RakNetInterface.java:66)                     │·
        at cn.nukkit.network.Network.processInterfaces(Network.java:75)                           │·
        at cn.nukkit.Server.tick(Server.java:1003)                                                │·
        at cn.nukkit.Server.tickProcessor(Server.java:800)                                        │·
        at cn.nukkit.Server.start(Server.java:779)                                                │·
        at cn.nukkit.Server.<init>(Server.java:470)                                               │·
        at cn.nukkit.Nukkit.main(Nukkit.java:68)``

**Comment by NycuRO** [comment]
@sero583 Thanks for notice. I'll try in next commit. It's with More Populators and fixes 👍 
Probably and decorators like Temple and others 👍 
I'll try.

**Comment by NycuRO** [comment]
I can't continue this project because Nukkit it's so outdated. 

Why i'm say that? 

Because i can't do what i want. I need port from Bukkit/Spigot all and in final get errors in gameplay , non aditional features and not use Nukkit.

Anyone can continue this project.

Port what you need from this: https://github.com/Creeperface01/MinecraftSource/tree/master/src/net/minecraft/world

And you can port and from there Structures:
https://github.com/Ad5001/BetterGen

I'm so sorry guys, but nukkit it's for developers. If i want one developer-soft i can use SteadFast2 or MiNET. 

@yescallop  don't update to 1.2, @Creeperface01 can't fixed nether, and packets != pm-mp. (More good).

In the last commit:

- I'm fixed obsolete drop.
- I'm optimized TNT (Minor), if you want one Major optimization, port this: https://github.com/pmmp/PocketMine-MP/commit/da83b879c09cb412dc3a6093920c1c415e682532
- I'm dezactivated Nether because don't work.

**Comment by NycuRO** [comment]
Please don't close this PR or other, you can push changes because it's tested and more optimized like normal.

New addons are good.

**Comment by NycuRO** [comment]
Now i'm going watch bugs and fix and i'm coming out from this.

**Comment by NycuRO** [comment]
P.S. If you want fix knockback, have there fix: https://github.com/pmmp/PocketMine-MP/commit/2f3c77c68a0d9e0fdfb0eb69bd44344f4c8d028d

**Comment by NycuRO** [comment]
image

"Ravines". @Creeperface01 You not ported correctly from Steadfast2-Java ;)

P.S. Last commit i added CactusPopulator again and DeadBush to Desert Biome.


NOTE: Populators are inccorect. All populators is generated with divide(4) by vector3.. "chunkX and chunkZ >> 4, for more easy".

@PikyCZ

File: src/main/java/cn/nukkit/entity/passive/EntitySquid.java [hunk 3]
```diff
@@ -40,15 +46,92 @@ public float getEyeHeight() {
...
+                    this.motionX = this.swimDirection.x * this.swimSpeed;
+                    this.motionY = this.swimDirection.y * this.swimSpeed;
+                    this.motionZ = this.swimDirection.z * this.swimSpeed;
+                }
+            } else {
+                this.swimDirection = this.generateRandomDirection();
+                this.swimSpeed = random.nextInt((50) + 100) / 2000;
+            }
+
+            // Much math there <3 (Yeah, i love math, @NycuRO)
```
Comment (NycuRO) [inline src/main/java/cn/nukkit/entity/passive/EntitySquid.java]: @larryTheCoder math <3

**Comment by NycuRO** [comment]
```java
20:35:33 [EMERGENCY] Exception happened while ticking server
20:35:33 [ALERT] java.lang.StackOverflowError
        at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2883)
        at com.google.common.cache.LocalCache.get(LocalCache.java:4129)
        at cn.nukkit.level.Level.getBlock(Level.java:1508)
        at cn.nukkit.level.Level.getBlock(Level.java:1497)
        at cn.nukkit.level.Level.getCollisionCubes(Level.java:1403)
        at cn.nukkit.entity.Entity.checkObstruction(Entity.java:926)
        at cn.nukkit.entity.item.EntityItem.tryChangeMovement(EntityItem.java:149)
        at cn.nukkit.entity.item.EntityItem.tryChangeMovement(EntityItem.java:150)
        at cn.nukkit.entity.item.EntityItem.tryChangeMovement(EntityItem.java:150)
        at cn.nukkit.entity.item.EntityItem.tryChangeMovement(EntityItem.java:150)
        at cn.nukkit.entity.item.EntityItem.tryChangeMovement(EntityItem.java:150)
        at cn.nukkit.entity.item.EntityItem.tryChangeMovement(EntityItem.java:150)
```

I fix tommorow. Good night all.

**Comment by sero583** [comment]
@NycuRO fly through sand biomes, you only see cactus with height 1, maybe add that most of them are 3-2 high

**Comment by NycuRO** [comment]
I know that @sero583 , but i don't have time now..

**Comment by NycuRO** [comment]
Ops, no.

Need re-add recipes.

Wait pls.

**Comment by SupremeMortal** [comment]
Fix the conflicts yourself. You just need to merge the upstream changes and edit the conflicts.

**Comment by NycuRO** [comment]
Ready.

**Comment by PeratX** [comment]
It has been a long road.

**Comment by NycuRO** [comment]
@PeratX You can accept as one commit.

**Comment by PeratX** [comment]
More tests needed.

**Comment by NycuRO** [comment]
I'm tested so much. More it's fixed.

You can test what you what.

Only one bug: When you put one squid, get error console spam, i don't know how fix it, pmmp credits for fix errors movement.

**Comment by sero583** [comment]
@NycuRO Well honestly I totally agree @PeratX, this is not a commit which you just merge after few tests, this enhancement should fix and dont bringt more bugs. By the way was this tested for longer time than some minutes, also is the code a performance killer or can be optimized to run as fast as possible, a comparison from before and this PR would be good

**Comment by PeratX** [comment]
As this is not a small enhancement, more tests should be done. Including performance test, generator tests and so on.

**Comment by sero583** [comment]
@NycuRO @PeratX Maybe a comparison PR vs now would be good?

**Comment by SupremeMortal** [comment]
I think you ought to close this if you've merged someone else branch into your pull request.

**Comment by NycuRO** [comment]
Wtf?

**Comment by NycuRO** [comment]
Omgggg!

**Pull request closed without merging**